File: apps/hermes/client/rs/examples/hermes_client.rs
use futures_util::{pin_mut, StreamExt};
use hermes_types::{AssetType, EncodingType};
use price_service_client::hermes_client::{HermesClient, HermesClientConfig, ParamOption};

const ENDPOINT: &str = "https://hermes.pyth.network";

#[tokio::main]
async fn main() {
    let config = HermesClientConfig::new(None);
    let connection = HermesClient::new(ENDPOINT, Some(config)).expect("Failed to construct client");

    let price_ids = vec!["e62df6c8b4a85fe1a67db44dc12de5db330f7ac66b72dc658afedf0f4a415b43"];

    // Get price feeds
    let price_feeds = connection
        .get_price_feeds(Some(ParamOption::new(
            Some("btc".to_string()),
            Some(AssetType::Crypto),
            None,
            None,
            None,
            None,
        )))
        .await
        .expect("Failed to get price feeds");
    println!("Price Feeds length: {}", price_feeds.len());

    // Laetst price updates
    let price_updates = connection
        .get_latest_price_updates(&price_ids, None)
        .await
        .expect("Failed to get latest price updates");
    println!(
        "Price updates binary data length: {}",
        price_updates.binary.data.len()
    );

    // Streaming price updates
    let price_updates_stream = connection.get_price_updates_stream(
        &price_ids,
        Some(ParamOption::new(
            None,
            None,
            Some(EncodingType::Hex),
            Some(true),
            Some(true),
            Some(true),
        )),
    );
    pin_mut!(price_updates_stream);

    while let Some(price_update_result) = price_updates_stream.next().await {
        match price_update_result {
            Ok(price_update) => {
                // Process the price update
                println!("Received price update: {:?}", price_update);
            }
            Err(e) => {
                eprintln!("Error receiving price update: {:?}", e);
            }
        }
    }
}

File: apps/hermes/client/js/src/examples/hermes-client.ts
import yargs from "yargs";
import { hideBin } from "yargs/helpers";
import type { PriceUpdate } from "../HermesClient";
import { HermesClient } from "../HermesClient";

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

const argv = yargs(hideBin(process.argv))
  .option("endpoint", {
    description:
      "Endpoint URL for the price service. e.g: https://endpoint/example",
    required: true,
    type: "string",
  })
  .option("price-ids", {
    description:
      "Space separated price feed ids (in hex without leading 0x) to fetch." +
      " e.g: f9c0172ba10dfa4d19088d...",
    required: true,
    type: "array",
  })
  .help()
  .alias("help", "h")
  .parserConfiguration({
    "parse-numbers": false,
  })
  .parseSync();

/**
 * Extracts the endpoint and basic authorization headers from a given URL string.
 *
 * @param {string} urlString - The URL string containing the endpoint and optional basic auth credentials.
 * @returns {{ endpoint: string; headers: HeadersInit }} An object containing the endpoint URL and headers.
 */
function extractBasicAuthorizationHeadersFromUrl(urlString: string): {
  endpoint: string;
  headers: HeadersInit;
} {
  const url = new URL(urlString);
  const headers: HeadersInit = {};

  if (url.username && url.password) {
    headers.Authorization = `Basic ${btoa(`${url.username}:${url.password}`)}`;
    url.username = "";
    url.password = "";
  }

  return { endpoint: url.toString(), headers };
}

async function run() {
  const { endpoint, headers } = extractBasicAuthorizationHeadersFromUrl(
    argv.endpoint,
  );
  const connection = new HermesClient(endpoint, { headers });

  const priceIds = argv.priceIds as string[];
  const _priceFeeds = await connection.getPriceFeeds({
    assetType: "crypto",
    query: "btc",
  });
  const _priceUpdates = await connection.getLatestPriceUpdates(priceIds);
  const eventSource = await connection.getPriceUpdatesStream(priceIds, {
    allowUnordered: false,
    benchmarksOnly: true,
    encoding: "hex",
    parsed: true,
  });

  eventSource.onmessage = (event: MessageEvent<string>) => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const _priceUpdate = JSON.parse(event.data) as PriceUpdate;
  };

  eventSource.onerror = (error: Event) => {
    eventSource.close();
  };

  await sleep(5000);
  eventSource.close();
}

run();
